### DISCUSSION TITLE
Switching from HBBFT to AlephBFT

### DISCUSSION THREAD

**Discussion by joschisan** [fedimint/fedimint#2086] [General]
## Motivation

The finalization latency of the HBBFT algorithm behaves roughly like t + l + (log(n - f) + log(f)) * l where t is the troughput, l is the network latency, n is the number of nodes and f is the number of faulty nodes for the following reasons:

* when we increase the thoughput the time to broadcast increases. As broadcast and async binary agreement are not parallelized this is quickly felt in terms of total round time.
* the broadcast requires 3 rounds to complete.
* when we run n instances of asynchronous binary agreement concurrently the expected number of rounds is log(n). However, as soon as we have one faulty node the n ABA instances cannot be executed concurrently but we have to wait for a threshold of ABAs to finish before we can make any progress on any of the ABAs corresponding to faulty nodes. This leads to the sudden jump in latency when going from zero faulty nodes to one.

From the equation we can see that, given a maximal acceptable transaction latency of 6 seconds or so we would have to carefully balance the federation size, faulty nodes, throughput and network latency. The latency tests by @elsirion indicate that this already a problem in many realistic deployment scenarios of fedimint. Therefore, federations with more throughput than 1 transaction per second or federations with more than seven guardians are not feasible as they can not provide the required time to finalisation in many network condtions. 

## Possible Solution

I have reason to believe that these tradeoffs can be practically removed by a different choice of BFT algorithm called Aleph BFT. In contrast to HBBFT, which explicitly trades latency for throughput, the algorithm has been designed with low latency in mind. The design of Aleph BFT removes all previously listed problems with HBBFT and, in theory, should enable fedimint to achieve low latency practically irrespective of throughput, federation size and number of faults. The theoretical expected transactions latency is 5.5 times the one way network latency of the n-f fastest correct nodes where f is the maximal number of possible faults. If the theoretical properties of the algorithm hold in practice switching to AlephBFT should provide reliable low latency even in federations with 100 tps and two dozen guardians.

There is good documentation available for the algorithm at https://github.com/Cardinal-Cryptography/AlephBFT which gives a good overview and the paper is available at https://arxiv.org/abs/1908.05156. 

## Verifying theoretical performance

There is an implementation available on crates.io which I used to generate the following latency test results. All tests were run with 100ms simulated one-way latency and the maximum possible number of faulty nodes. 

| Federation Size | Faulty Nodes | Avg. Time to Confirmation (ms) | Max. Time to Confirmation (ms) |
|-----------------|--------------|--------------------------------|--------------------------------|
| 4               | 1            | 473                            | 511                            |
| 7               | 2            | 496                            | 589                            |
| 10              | 3            | 511                            | 602                            |
| 13              | 4            | 527                            | 661                            |
| 16              | 5            | 568                            | 633                            |
| 19              | 6            | 622                            | 827                            |
| 22              | 7            | 659                            | 756                            |

The results are consistent with the theoretical performance characteristics claimed in the original white paper. As you can see the algorithms scales well with respect to federation size.

## Potential Problems

There are two potential problems with this algorithm / its implementation compared to HBBFT. However, it seems to me that both can solved.

### Asynchronous Liveness

The implemented algorithm differs from the paper as it does not provide common randomness hence it cannot guarantee asynchronous liveness against a powerful network advisary. However, the authors argue that such an attack has not been demonstrated in practice and that one could later add the option for randomness to the algorithm. Furthermore, they argue that it makes more sense to add this to a well tested algorithm and only if necessary as it introduces substantial complexity to security critical code. I find this a convincing argument as the ability to guarantee asynchronous liveness is much less important in practice than safety. Please note that the algorithm still will work just fine in bad network conditions without a network adversary.

### Sessions

While the algorithm runs its RAM consumption grows linearly with the number of processed transactions as the communication history is recorded. This history needs to be kept in RAM for consensus to remain fast as it is relevant to consensus until the current session is closed. A session is a preset number of rounds (default) 3000 and the implementation requires the user to sequentially create new sessions and switch between them. Switching between session allows to discard the communication history and message queues. The  implementation specifically leaves it to the users to handle the switching in a application specific way.

Switching to the next session would result in 5.5 *latency of downtime depending on the implementation. However, in theory, we should be able to overlap sessions by a second or so to proceed without any downtime. Depending on usage and RAM the switch should only happen every couple hours to days. 

## Crash Recovery

When the algorithm has completed the preset number of rounds it has output all transactions for this session. We then can hash all the transactions from this session and keep the session open for enough nodes to complete the preset number of rounds as well, so we can create a n - f threshold signature of the session number and the hash. After we have created this signature and saved it to disk we can safely close the session and discard the message queues;  lagging nodes can simply download the entire sessions transactions from us and verify the signature. However, we would keep the session open for another minute or so such as to not trigger a download of the entire session unnecessarily in case a node is not far behind.

In case a node crashed it could simply request the threshold signed hashes for the missed sessions, download the corresponding transactions from its peers, update its own state accordingly and join the current session. A crashed node does not have to wait for the next session to join as alephbft can backup all data required to rejoin a session to disk. The backup only requires a simple filesystem. If we rejoin a session we iterate through all its transactions from the start again, thereby not missing any transaction from the crash. Therefore the system can recover automatically even if all nodes crash simultanously without losing any transactions or creating inconsistent state.

**Comment by elsirion** [comment]
Some comments/questions while reading:

* What do we need to ensure for sessions switches to be secure, I remember you talking about signing of system state at some point. Let's verify if we already meet the requirements.
* How would overlapping sessions work?

> Find a way for realistic latency testing of Fedimint; possibly expose an endpoint for telemetry data.

We can likely use `tracing` for this, probably using [`tracing_timing`](https://docs.rs/tracing-timing/latest/tracing_timing/).

For crash recovery it might be good to involve @jkitman since he designed the current system. 

> A crashed not does not have to wait for the next session to join as alephbft can backup all data required to rejoin to disk.

This would mean saving any msg before sending to disk and keeping it at least till it's ACKed by the peer? @dpc had a similar idea some time ago.

**Reply by jkitman** [reply]
We should consider how robust the federation is in the case of `f+1` nodes crashing.  If we broadcast a `1 BTC` peg-out but haven't stored the spent e-cash in the database yet then we are in big trouble if we crash.  In HBBFT this is somewhat handled by the fact that peg-outs or LN payments take multiple epochs, and each epoch is written to disk before the next one commences.

Since AlephBFT keeps sessions in-memory, we may need to introduce an additional consensus message that confirms a transaction was written to disk before we go on to process the payment.

**Reply by joschisan** [reply]
> What do we need to ensure for sessions switches to be secure, I remember you talking about signing of system state at some point.

Instead of signing the system state it would probably make more sense to threshold sign the hash of all transactions in a session. This later allows lagging nodes to download the result of a missed session.

> How would overlapping sessions work?

We could feed the transactions in two sessions at the same time. Then, when the first session terminates we can immediately output the transactions from the second session. This prevents delay in case a transaction that is input to the first session right before the switch does not get output before the first session terminates.

> This would mean saving any msg before sending to disk and keeping it at least till it's ACKed by the peer?

I think the implementation just saves the current DAG to disk, so all received and created units.

**Reply by timorl** [reply]
Oh I see you figured out the parallel sessions here already, so I guess I didn't have to describe it in the other issue.

Anyway, I wanted to chip in with a small correction – currently ABFT _does not_ save the _whole_ dag to disk, so it's not quite safe against crashes. There are reasons to expect it to survive anyway, but in sufficiently Byzantine situations the saved state might not be sufficient to rejoin a session. However, we are currently in the process of implementing that, the next release (<=2months) should have full support for rejoining session after crashes in any BFT situation (as long as on-disk data has not been lost ofc).
